### ISSUE TITLE
"Rewrap on resize" causing crashes

### ISSUE THREAD

**Issue by clayne** [mintty/mintty#1150]
Early apologies for the complete lack of actual technical info but it seems the new rewrap feature has some undefined behavior going on somewhere. I've noticed about 5% of the time on resizing windows mintty will just flat out crash (although I've seen it hang once as well). This _only_ started after enabling the new rewrap functionality so my gut tells me the feature is directly involved.

I suppose I could try and capture it in a debugger, but it's infrequent enough that it may take a while.

**Comment by Biswa96** [comment]
Can you find any pattern of those crashes? Like with any specific programs or in any specific environment etc.

**Comment by clayne** [comment]
I'm attempting to see if I can reproduce it some how but it's almost always just a general shell being open with some leftover scrollback still in it and attempting to resize the window as part of general making space on my screen. It's not really a specific program that's running.

One other thing I did notice is that the speed of resizing a window is directly correlated to the amount of scrollback lines present in the term (with this option enabled). I'm beginning to wonder if large amounts of scrollback are lending itself to some kind of race somewhere as I usually run with around 100k lines of scrollback limit.

**Comment by mintty** [comment]
Please use the attached debug version. If it crashes, zip and upload file /tmp/mtlog.
[mintty-debug.zip](https://github.com/mintty/mintty/files/8382582/mintty-debug.zip)

**Comment by clayne** [comment]
I have multiple terminals running so the log is going to show other non-involved ones, but I just had a resize completely hang one:

![image](https://user-images.githubusercontent.com/2467434/161112951-7bc6cc14-830b-43d5-8862-dd2ae4e5c725.png)

This mintty is pid 3269...:

```
clayne@opteron:~ $ ps -efW | grep mintty
  clayne    3629       1 ?        09:18:22 /usr/bin/mintty-debug
  clayne    3712       1 ?        10:16:46 /usr/bin/mintty-debug
clayne@opteron:~ $ cat /tmp/mtlog | grep 3629
[3629] resize 82 119 @0 (0/0 @0)
[3629] resize 102 184 @0 (354/100000 @354)
[3629] resize 107 185 @0 (2362/100000 @2362)
```

**Comment by mintty** [comment]
I cannot reproduce such size, can you demonstrate it with smaller screen size?
Also, please add a log `mintty -l .log`, and your configuration.

**Comment by clayne** [comment]
Sure. It happens with smaller sizes as well, just any resize with what seems like a particular combination of scrollback depth (or potentially content in scrollback) and the right timing. It feels really race-like to me.

**Comment by mintty** [comment]
Thousands of random size changes did not reveal a specific crash condition but some code analysis did.
Test case (for an 80-column window): [xwrapcancel.txt](https://github.com/mintty/mintty/files/8495874/xwrapcancel.txt)
 Fix uploaded, and also fixed a memory leak that became apparent.

**Comment by mintty** [comment]
Fixed in 3.6.1.
